### ISSUE TITLE
Import Existing Movies On Disk - needs 'add all' function

### ISSUE THREAD

**Issue by luxferro** [Radarr/Radarr#150]
Movies>Add Movie>Import Existing Movies On Disk

When you point it to your existing movie library, it will scan folders, find metadata, and leave you with a long list, with the only option of adding each item individually. Not fun if you have 100's and 1000's of movies.

**Comment by fedoranimus** [comment]
I fundamentally disagree. There is no way to guarantee that the match is correct. If the user selects "Add All", then the mismatched file will be misnamed and extremely difficult to tease this out.

If we were to do this, we need to inform the user of this possibility.

**Comment by onedr0p** [comment]
@fedoranimus I agreed, maybe leave this issue here for other people to see our conclusion.

**Comment by galli-leo** [comment]
@fedoranimus @onedr0p We could make a dialog where the folder name and the found movie is displayed. That way it's quicker to find any wrong ones, delete them and then add all.

**Comment by ShanDestromp** [comment]
@fedoranimus I personally disagree.  Last night I added a mere 600 movies to Radarr and it took me well over an hour to do so because of severe lag after adding the first hundred or so.  The lag got "better" once I closed some of the other radarr tabs I had open.  Please note this lag wasn't an issue with my personal machine as Chrome wasn't using much RAM and no other (non radarr) tabs had this lag.  Furthermore when I closed the extra radarr tabs the memory usage used by the radar process on my server machine halved.  It is still however incredibly high compared to my sonarr process.

Out of those 600 movies only one was incorrectly identified and I'm not sure why.

The particular movie was Audition (imdb tt0235198); it was identified as imdb tt2246575; and this is despite my collection being formatted as NAME (year) [resolution]/name (year).ext.  Furthermore this is an existing collection with valid nfo files which already contain the imdb id etc.

I'm not saying the "add all" button needs to be the default action, perhaps have it at the very bottom of the existing page allowing users to tweak individual movies *prior* to "add all"; but it definitely would have been nice to bulk add them instead of clicking the "+"; wait for the page to catch up and then repeating 600 times.

**Comment by galli-leo** [comment]
@fedoranimus @onedr0p I thought of something like this: ![Add Multiple](http://i.imgur.com/naKVzVM.png)

(Just a quick mockup)

**Comment by onedr0p** [comment]
@galli-leo but what happens when your movie is ambiguous?

**Comment by donldmn** [comment]
I agree that an "Add All" button would be nice for importing a large library. I also agree that it could seriously mess a library up.

The quick mockup above seems great because it look like it allows you to deselect a movie that is falsely identified.

**Comment by galli-leo** [comment]
And yeah, we can also add a column with the imdbid where you can manually change it.

**Comment by StevenChorkley** [comment]
Import-All is definitely an important feature, I've used it for years with CouchPotato. I had metadata included during post-processing, so I've never had an incorrect match.

What about: 
All low-confidence/unknown matches being displayed to the user once the import has taken place, then allow the selection/search for the correct match.

**Comment by sch1308** [comment]
it looks like a great idea just from tinkering around for a few minutes (coming from a long time sonarr user) but without an add all i'd never get thru adding my 5,000 movies from CP. 
![image](https://cloud.githubusercontent.com/assets/7896067/21864662/bfc6b5ac-d810-11e6-8975-dc19fdd0aece.png)

**Comment by simonk83** [comment]
Should "Monitored" default to No, rather than Yes?  If I'm importing my existing movies on disk, I obviously already have them :D

**Comment by sch1308** [comment]
I'd argue monitored should be yes by default unless the qualifications are
satisfied in settings (1080p vs Blu-ray qualities ) just like couch potato

**Comment by fedoranimus** [comment]
@galli-leo Okay, we could do something like that.

**Comment by simonk83** [comment]
> I'd argue monitored should be yes by default unless the qualifications are
satisfied in settings (1080p vs Blu-ray qualities ) just like couch potato

No biggy, I forgot that if you change monitored to No once, it changes it for every other movie as well, so it's not a big deal either way :)

**Comment by wcomartin** [comment]
Should the "Already Added" movies be hidden from the Import Existing, or at least the folders, and perhaps a way to add a second version of the already added movies if the folder is different?

**Comment by simonk83** [comment]
Yep that'd make things easier. Even just a show/hide toggle for already
added.

**Comment by onedr0p** [comment]
@galli-leo @bigbasec suggestion to `hide existing items in the database when Importing from Disk` is a good idea.

**Comment by SpencerCarstens** [comment]
What about showing a list with checkboxes that allows you to see the movie dir and then the matched name. So you can visually scan it and then fire it off.

```
[   ] -- Location: "/location/movies/{ movie_title_folder }/{ movie_title.mkv }" 
      -- Matched: { matched_movie }
[ X ] -- Location: "/location/movies/{ movie_title_folder }/{ movie_title.mkv }" 
      -- Matched: { matched_movie }
[ X ] -- Location: "/location/movies/{ movie_title_folder }/{ movie_title.mkv }" 
      -- Matched: { matched_movie }
[ X ] -- Location: "/location/movies/{ movie_title_folder }/{ movie_title.mkv }" 
      -- Matched: { matched_movie }

[ Add All Selected ]  
```

**Comment by Wreave** [comment]
I like that

**Comment by galli-leo** [comment]
That's what I had in mind with my quick mockup

**Comment by estrellagus** [comment]
this is a must add before can fully use;  Those with thousands of movies just takes too long to add

**Comment by onedr0p** [comment]
@estrellagus, until then you can use it to track unreleased movies or movies in theaters :)

**Comment by nickk024** [comment]
Love this add. Agree on definite need for this feature.

**Comment by benjmd** [comment]
While we're bulk adding, we should also be able say the quality of what we are importing (matching available profiles?)
I manually added all of my existing movies (yay) but it thinks they're all xxTV variants, not 1080p-BluRay/whatever they actually are.
Is there a way to assign a 'downloaded' quality type?

![downloaded quality](https://cloud.githubusercontent.com/assets/18607108/22061136/8c8d8194-ddc3-11e6-8f8c-bebe8a5f1e05.PNG)

For some reason it's also snatching movies that are unmonitored, but that's a problem for another 'issue'

**Comment by PhAzE-Variance** [comment]
How about an option to tick the movies you want to add, then the add all button adds the movies that are selected. This way users can check through and quickly tick off correctly matched movies rather than having to wait for the movie to be added one at a time.

**Comment by dawnerd** [comment]
The big problem I see right now is the page will crash chrome before it finishes loading a really large set of movies.

**Comment by wesleykirkland** [comment]
I would like to see this feature, its a feature that CouchPotato has had. Going through and adding my movies manually Radarr has a 98% match right. It always find the correct match except when the naming is off.

**Comment by MrMxyzptlk** [comment]
Radarr looks like a great competitor to CouchPotato, but I won't be able to use this, until add all is coded.
Thanks for the great work!

**Comment by JohnnyT009** [comment]
Thank you for adding my topic to this. I would even be ok if it showed 200 per page with match results and tiny art. You could add all those and then move onto the next. Checkboxes everywhere :-)

**Comment by PhAzE-Variance** [comment]
Actually that's a good point. Perhaps instead of an infinite loading of movies on one page, the list needs to be separated into multiple pages for load reducing purposes.

**Comment by iairj84** [comment]
I am Okay with having to click a box for each movie, but I am not good with it defaulting to "Monitor" the existing library. I would rather it default to not monitor and if necessary I can go back in and change it. If not I can see it downloading several hundred films because they're old DVD rips instead of HD.

**Comment by seriosbrad** [comment]
I would absolutely love a couple of options like:

- Add All and Monitor
- Add All and Not Monitor

**Comment by drafie** [comment]
Love whats going on with Radarr, but 7,000+ movies one by one is a no go.  This is a must have feature, for folks with large libraries before we can use it.

**Comment by danejaho** [comment]
I agree with drafie. This is a fatal flaw in Radarr's current coding. You have to assume, due to the length of time Couchpotato has been around, that the people using Radarr are making the switch from that program to this one. As such you must also assume that these people already have a movie collection. Few people are going to want to take the time to manually add each and every movie they have into Radarr.

**Comment by galli-leo** [comment]
@danejaho For that we have a Couchpotato import on the latest nightly branch :)
Anyways, you can still start using Radarr for upcoming movies, as I don't see any necessity in adding your existing collection right now.

**Comment by simonk83** [comment]
I mean as Gallileo said, you can now import your wanted movies into Radarr,
but actually having your existing movie collection in Radarr really serves
no purpose unless you plan on upgrading every movie you have.

I agree it'll be a handy feature to import all, but it's hardly the be all
end all. I've been using Radarr happily since it was started and haven't
bothered adding my existing library, just my wanted list.

**Comment by danejaho** [comment]
It's not so much about wanting to update each movie, it's more because when you have a large movie collection it's hard to keep track of what you have or don't have. I've already found Radarr useful (as I went though and added most of my movie collection manually) because I searched for a movie thinking I didn't have it yet and Radarr let me know it was already in my collection and in 1080p so there was no need to look for a new or better copy.

That's great news that the nightly has a couch import! Good work guys.

**Comment by simonk83** [comment]
Yep that's fair enough. They are working on it though so it'll get there :)

**Comment by Nodja** [comment]
I managed to import my collection of about 1000 movies. I wanted to do it because I wanted to make sure my movies are upgraded.

Use at your own risk:
1. Open the import movies from disk page and select the folder, make sure you have all the movies you want to import showing on the webpage.

2. Open the javascript console while on that page (on Chrome  ctrl+shift+I)

3. Copy paste these lines one at a time and then press enter.
 ```
 $('#add-movies-workspace').find('.x-monitor').val('none')
 $('#add-movies-workspace').find('.btn.x-add').click()`
 ```

The first line should set all monitored to no, the second line should click on all the add buttons.

Get ready to have radarr be sluggish for a while :) Good luck.

**Comment by proxenus** [comment]
Just wanted to mention another reason for importing existing movies - If you are also using a watch list for new downloads it should prevent downloading a movie already from the watch list if it's already in your library.

**Comment by crafty35a** [comment]
@galli-leo, thanks for your work on this so far in the develop branch! With the latest build, bulk import seems to be working great, except the added movies don't have the local files associated, so they come in as "not downloaded." Just a heads up, in case you weren't aware.

**Comment by shushry** [comment]
First, it's GREAT to see this branch / project get created.   And I +1 the "add all" feature.  At some point you have to give some trust to the end user that they know what they are doing.   Folks moving to this will need to add large collections, and you want that imported BEFORE you add any watch/wanted list, to, as stated above, avoid duplicates.

**Comment by mommalongnips** [comment]
Just to jump on the band wagon, I hate couchpotato with a passion and am very excited to make the switch to Radarr. However, I won't be making the switch until this feature exists :) Subscribed!

**Comment by drafie** [comment]
@galli-leo thanks for your work on this.  I see it here in the newest developer build.  I have a network share that I have not imported any movies from yet.   So, I tested it with that one. 

It found 6,174 movies - which is correct.   It paged them across 412 pages. 

http://prntscr.com/e5sa61

I clicked "Add All" http://prntscr.com/e5sck3, nothing happened.   

So I checked all of the ones visible on the page and clicked "Add All", nothing happened. 

I then clicked "Add Selected" and it did import those 15 movies on the screen.   http://prntscr.com/e5sd2l 

Not sure if its something specific to me or if "Add All" is not working properly yet.

**Comment by crafty35a** [comment]
@drafie For the 15 that you imported, did they come in correctly with a reference to the file on disk? Or are were they imported with no file, and marked as "Not downloaded"? Last I checked, this is what was happening. Of course, this could just be that the code for this feature isn't quite ready yet.

**Comment by drafie** [comment]
@crafty35a I spot checked one - because I don't know what the other 14 were.  It does have a reference to the file on disk and it did scan it to determine that it is 1080p and it says "Downloaded".

**Comment by galli-leo** [comment]
@drafie Add all only works if you went through all pages. This is to make sure you don't accidentally add a wrongly matched movie, which would be very hard to detect afterwards.

**Comment by galli-leo** [comment]
I also fixed add all something like 5-6 hours ago, maybe you don't have that fix yet?

**Comment by crafty35a** [comment]
@drafie That's great news! Thanks for your work on this, @galli-leo.

**Comment by JohnnyT009** [comment]
Any way for it to also scan subfolders? I have movies in collections "fast and furious collection" "Ace Ventura Collection" and it will fail to look inside of those when I bulk add to also add those movies. It just tells me "Sorry. We Couldn't Find Any Movies Matching 'Ace Ventura Collection'"

**Comment by galli-leo** [comment]
@JohnnyT009 This would best be handled in a separate issue (there already is one). But I don't think we are going to support importing that in any way. To much unknown variables.

**Comment by drafie** [comment]
@galli-leo if I were to click through all 471 pages, the problem is that each page click is taking quite a long time.  The spinner goes and goes before it switches.   I think this is due to the number of movies it is importing. 

I think you should let end-users assume the risk here and not require them to click through hundreds of pages before being able to add-all.   While, better than nothing - its not particularly usable or friendly.   If you still feel very strongly that users must be forced to review everything then I suggest setting the paging mechanism to 100 per page so we can quickly scan down the page.    Basically, we're all probably looking to make sure the folder name on disk matches the suggested movie name.    

Thanks!

**Comment by luxferro** [comment]
Another way of preventing mishaps, is to prevent the 'Import' function from renaming files. Then it's easy to catch mismatched imports. 

Limit the write functions (renaming, ect) so they only happen when it snags a new file, and moves it from the download folder to your media folder.

**Comment by drafie** [comment]
as an FYI - its hard for me to test this functionality, because I am plagued by this Sonarr issue here: https://forums.sonarr.tv/t/import-existing-series-on-disk-gets-stuck-quite-often/8136  

Hopefully some others can give this a thorough testing.

**Comment by PhAzE-Variance** [comment]
Alternatively, implementing issue #355 might help. This way if you find somethign incorrectly matched after the fact, you can fix it without having to delete and re-import the movie.

**Comment by galli-leo** [comment]
@drafie We can add an option to set the page size to what you desire. However, keep in mind that the larger the page size, the longer it will take a page to load (since it has to search TMDB for each movie on that page).

**Comment by drafie** [comment]
@galli-leo that would be a big help.   I did import 1,200 on a single page, using the existing Import From Disk plus javascript console commands @Nodja provided.

**Comment by MrMxyzptlk** [comment]
Love this feature, but it's not really working correctly for me anyway.
I have 1953 movies in CouchPotato, and I setup a list in Radarr to import all movies, unmonitor, and when I fetch this list only 2 movies show up each time...the same 2 movies.

All of my movies are in the same folder with subfolders as the movie name / date, and have everyone permissions?

**Comment by galli-leo** [comment]
@MrMxyzpltk Only not yet downloaded movies from CP show up

**Comment by onedr0p** [comment]
@MrMxyzptlk use the soon to be release new bulk import feature when you want to import existng movies.

**Comment by MrMxyzptlk** [comment]
OK, thanks for the replies!

**Comment by crafty35a** [comment]
@galli-leo Any idea why I'm getting "No movies found in folder D:\Movies\. Have you already added all of them?" now when I select my movies folder for Bulk Import? Bulk Import was successfully reading movies from this folder a couple days ago, and it's still able to read movies from other locations.

**Comment by delphiactual** [comment]
bulk import should not fail with `Invalid request Validation failed: 
 -- Path is already configured as a root folder`

**Comment by drafie** [comment]
@delphiactual that is similar to the problem I have.   Do you see anything under Recent Folders or do you just see bouncing balls?

**Comment by galli-leo** [comment]
@delphiactual @drafie Can you post a full debug log? Sonarr has the same problem, idk what they did to fix that.

**Comment by drafie** [comment]
@galli-leo here you go.   I did see there is an issue with the movie Sing (2016).   I don't know that the logs are particularly helpful.    90% of the time, it doesn't show anything under Recent Folders.   I also discovered that when this problem occurs, if I go to Add Movie and pick a new movie then the Add Path drop down is empty, only offering to add a new path (which returns the root path already configured error). 

[radarr.txt](https://github.com/Radarr/Radarr/files/762028/radarr.txt)

**Comment by galli-leo** [comment]
@drafie If this happens again, can you open the chrome dev console and post all errors from there?

**Comment by drafie** [comment]
@galli-leo 

Sure.  Here is what I see visually.   http://prntscr.com/e6ajjj 

vendor.js:9785 Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check https://xhr.spec.whatwg.org/.

send @ vendor.js:9785
main.js:37 starting application
main.js:201 starting signalR
poster-250.jpg Failed to load resource: the server responded with a status of 404 (Not Found)
poster-250.jpg Failed to load resource: the server responded with a status of 404 (Not Found)
298win7vm2/:1 Error in event handler for (unknown): TypeError: Cannot read property 'tab' of undefined
    at chrome-extension://kamiaabbnpncbobdiblhfogafnoefkae/contentscript.js:38:28
AddMoviesView.js:29 Object__proto__: Object
poster-250.jpg Failed to load resource: the server responded with a status of 404 (Not Found)
SignalRBroadcaster.js:32 SignalR: [connecting]
SignalRBroadcaster.js:32 SignalR: [connected]
3backbone.signalr.mixin.js:15 sync received, re-fetching collection
manifest.json:1 GET http://win7vm2:7878/Content/Images/favicon/manifest.json 500 (Internal Server Error)

**Comment by galli-leo** [comment]
@drafie Can you hop on our discord so we don't clutter this issue here?

**Comment by drafie** [comment]
@galli-leo is there a page that can be called directly, rather than loading it into an modal or iframe, to see the Recent Folders dialog?    The debugging probably needs to happen on that particular page/php.

Yes, I can jump on discord.  Will do.

**Comment by fedoranimus** [comment]
This feature has been merge. Please open specific issues if support is required.

**Comment by djetaine** [comment]
I'm running this on windows with build 0.2.0.299
I have about 1400 movies on one particular folder and by the time that radarr gets to loading the full list, chrome runs out of memory and crashes. I see some other people on here mentioning tabs, I don't have any. It just loads the whole list on one page. 
During a bulk import would it be possible to not load the movie posters? This would help cut down on the memory issue.

**Comment by galli-leo** [comment]
@djetaine Bulk import is not in .299 You have to go on the latest nightly

**Comment by mommalongnips** [comment]
I like @djetaine suggestion. To add to it, maybe make it so when you hover over an item, then it loads the cover and such so we can still easily verify using a cover?

**Comment by galli-leo** [comment]
@mommalongnips this is already the case in the bulk import on the nightly branch

**Comment by stempest** [comment]
I'm adding 3600 movies at the moment... Radarr get's a memory leak and I have to restart it to resolve. Should implement a process similar to Plex where it just guesses... I've been at this for 3 hours now and am 1000 movies down with a lot more to go.

**Comment by CyberMew** [comment]
Still having problem with long loading times and unable to set unmonitored and view 1000 and import all by default. It's buggy. I have 900+ but it says 1400+ selected and I have views set to 500 per page. I want to set unmonitored all by default so that I can go in later and modify each if I need be but there's no option for that too. Also, I have some folders with it couldn't detect the mkv file inside, how would I go about manually pointing to it?
